<a href="https://colab.research.google.com/github/ChronosXV/Metodologia_grupo2/blob/Feacture%2FMetodologia_g2/Metodologia_1_Grupo_2_Validado.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Nombre del proyecto:


## Integrantes

## Instalaciones requeridas

In [12]:
!pip install duckdb -q

## Importaciones de Librerias.

In [14]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import duckdb
import gc

from typing import List, Dict, Optional, Any
from google.colab import drive

#Codificacion

In [15]:
# Acceso al Drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [16]:
ruta_carpeta = "/content/drive/MyDrive/Datos_csv"

In [ ]:
# archivos_parquet = [
#     archivo
#     for archivo in os.listdir(ruta_carpeta)
#     if archivo.endswith('.parquet')
# ]

# for archivo in archivos_parquet:
#     print(archivo)

p_historico_2018.parquet
p_historico_2021.parquet
p_historico_2020.parquet
p_historico_2019.parquet
p_BaseUnificadaEstaciones.parquet
p_historico_2017.parquet
p_historico_2018_limpio.parquet


In [ ]:
# # Conversion de Archivos csv a parque

# ruta_carpeta = "/content/drive/MyDrive/Datos_csv"

# # Ensure the directory exists
# if not os.path.exists(ruta_carpeta):
#     print(f"Error: La carpeta '{ruta_carpeta}' no existe.")
# else:
#     print(f"Archivos encontrados en '{ruta_carpeta}': {os.listdir(ruta_carpeta)}")
#     print("\nIniciando conversión de archivos CSV a Parquet...")

#     for archivo in os.listdir(ruta_carpeta):
#         if archivo.endswith('.csv'):
#             ruta_completa_csv = os.path.join(ruta_carpeta, archivo)
#             nombre_base = archivo.replace('.csv', '')
#             ruta_completa_parquet = os.path.join(ruta_carpeta, f'p_{nombre_base}.parquet')

#             try:
#                 df = None
#                 # Special handling for 'historico_2021.csv' with semicolon delimiter
#                 if archivo == 'historico_2021.csv':
#                     print(f"Intentando leer '{archivo}' con delimitador ';' y codificación 'latin1'...")
#                     df = pd.read_csv(ruta_completa_csv, sep=';', encoding='latin1')
#                 else:
#                     try:
#                         df = pd.read_csv(ruta_completa_csv, encoding='utf-8')
#                     except UnicodeDecodeError:
#                         print(f"El archivo '{archivo}' no se pudo leer con UTF-8. Intentando con 'latin1'...")
#                         df = pd.read_csv(ruta_completa_csv, encoding='latin1')

#                 if df is not None:
#                     df.to_parquet(ruta_completa_parquet, index=False)
#                     print(f"'{archivo}' convertido y guardado como '{os.path.basename(ruta_completa_parquet)}'")

#             except Exception as e:
#                 print(f"Error al procesar '{archivo}': {e}")

#     print("\nProceso de conversión a Parquet finalizado.")

Archivos encontrados en '/content/drive/MyDrive/Datos_csv': ['historico_2018.csv', 'historico_2021.csv', 'historico_2020.csv', 'historico_2019.csv', 'BaseUnificadaEstaciones.csv', 'historico_2017.csv']

Iniciando conversión de archivos CSV a Parquet...
'historico_2018.csv' convertido y guardado como 'p_historico_2018.parquet'
Intentando leer 'historico_2021.csv' con delimitador ';' y codificación 'latin1'...
'historico_2021.csv' convertido y guardado como 'p_historico_2021.parquet'
'historico_2020.csv' convertido y guardado como 'p_historico_2020.parquet'
'historico_2019.csv' convertido y guardado como 'p_historico_2019.parquet'
'BaseUnificadaEstaciones.csv' convertido y guardado como 'p_BaseUnificadaEstaciones.parquet'
El archivo 'historico_2017.csv' no se pudo leer con UTF-8. Intentando con 'latin1'...
'historico_2017.csv' convertido y guardado como 'p_historico_2017.parquet'

Proceso de conversión a Parquet finalizado.


In [ ]:
# # Creacion de dataSet por cada archivo parquet
# ruta_carpeta = "/content/drive/MyDrive/Datos_csv"

# dataframes_parquet = {}

# # Obtener la lista de archivos en la carpeta
# archivos_en_carpeta = os.listdir(ruta_carpeta)

# print("Iniciando carga de archivos Parquet a DataFrames...")
# for archivo in archivos_en_carpeta:
#     if archivo.endswith('.parquet'):
#         ruta_completa_archivo = os.path.join(ruta_carpeta, archivo)
#         try:
#             df = pd.read_parquet(ruta_completa_archivo)
#             dataframes_parquet[archivo] = df
#             print(f"DataFrame creado para '{archivo}'. Filas: {df.shape[0]}, Columnas: {df.shape[1]}")
#         except Exception as e:
#             print(f"Error al leer '{archivo}': {e}")

# print("\nDataFrames de Parquet disponibles (nombre de archivo: DataFrame):")
# for nombre_archivo, df in dataframes_parquet.items():
#     print(f"- {nombre_archivo}: {type(df)}")

# # Crear variables globales con el formato df_Nombrearchivo
# print("\nCreando variables de DataFrame individuales:")
# for nombre_archivo, df in dataframes_parquet.items():
#     # Quitar el prefijo 'p_' y la extensión '.parquet' para el nombre de la variable
#     nombre_base = nombre_archivo.replace('p_', '').replace('.parquet', '')
#     nombre_variable = f'df_{nombre_base}'
#     globals()[nombre_variable] = df
#     print(f"- Variable '{nombre_variable}' creada.")

Iniciando carga de archivos Parquet a DataFrames...
DataFrame creado para 'p_historico_2018.parquet'. Filas: 12136454, Columnas: 11
DataFrame creado para 'p_historico_2021.parquet'. Filas: 8071680, Columnas: 11
DataFrame creado para 'p_historico_2020.parquet'. Filas: 5781006, Columnas: 10
DataFrame creado para 'p_historico_2019.parquet'. Filas: 12662343, Columnas: 11
DataFrame creado para 'p_BaseUnificadaEstaciones.parquet'. Filas: 822363, Columnas: 6
DataFrame creado para 'p_historico_2017.parquet'. Filas: 11938476, Columnas: 12

DataFrames de Parquet disponibles (nombre de archivo: DataFrame):
- p_historico_2018.parquet: <class 'pandas.core.frame.DataFrame'>
- p_historico_2021.parquet: <class 'pandas.core.frame.DataFrame'>
- p_historico_2020.parquet: <class 'pandas.core.frame.DataFrame'>
- p_historico_2019.parquet: <class 'pandas.core.frame.DataFrame'>
- p_BaseUnificadaEstaciones.parquet: <class 'pandas.core.frame.DataFrame'>
- p_historico_2017.parquet: <class 'pandas.core.frame.Data

## Etapa 1: EDA

In [ ]:
# df_historico_2017.head()

,V1,FECHA,DESDE,HASTA,LINEA,MOLINETE,ID_ESTACION,ESTACION,PAX_PAGOS,PAX_PASES_PAGOS,PAX_FREQ,TOTAL
0,1,01/01/2017,08:00:00,08:15:00,D,LINEA_D_9JULIO_S_TURN02,6,9 DE JULIO,1,0,0,1
1,2,01/01/2017,08:00:00,08:15:00,D,LINEA_D_9JULIO_N_TURN02,6,9 DE JULIO,1,0,0,1
2,3,01/01/2017,08:00:00,08:15:00,D,LINEA_D_9JULIO_S_TURN01,6,9 DE JULIO,1,0,0,1
3,4,01/01/2017,08:15:00,08:30:00,D,LINEA_D_9JULIO_S_TURN02,6,9 DE JULIO,1,0,0,1
4,5,01/01/2017,08:15:00,08:30:00,D,LINEA_D_9JULIO_S_TURN01,6,9 DE JULIO,2,0,0,2


In [ ]:
# df_historico_2018.head()

,fecha,desde,hasta,linea,molinete,estacion,pax_pagos,pax_pases_pagos,pax_franq,total,periodo
0,2018-01-01,08:00:00,08:15:00,LineaA,LineaA_CBarros_S_Turn01,Castro Barros,1.0,0.0,0.0,1.0,201801
1,2018-01-01,08:00:00,08:15:00,LineaA,LineaA_Lima_S_Turn03,Lima,4.0,0.0,0.0,4.0,201801
2,2018-01-01,08:00:00,08:15:00,LineaA,LineaA_Pasco_Turn01,Pasco,1.0,0.0,0.0,1.0,201801
3,2018-01-01,08:00:00,08:15:00,LineaA,LineaA_Peru_S_Turn01,Peru,4.0,0.0,0.0,4.0,201801
4,2018-01-01,08:00:00,08:15:00,LineaA,LineaA_PJunta_S_Turn02,Primera Junta,2.0,0.0,0.0,2.0,201801


In [ ]:
# df_historico_2021.head()


,periodo,FECHA,DESDE,HASTA,LINEA,MOLINETE,ESTACION,pax_pagos,pax_pases_pagos,pax_franq,pax_TOTAL
0,2021,1/1/2021,08:00:00,08:15:00,LineaA,LineaA_Flores_Este_Turn02,Flores,1,0,0,1
1,2021,1/1/2021,08:00:00,08:15:00,LineaA,LineaA_SanPedrito_Oeste_Turn06,San Pedrito,0,0,2,2
2,2021,1/1/2021,08:00:00,08:15:00,LineaB,LineaB_Alem_N_Turn01,Leandro N. Alem,1,0,0,1
3,2021,1/1/2021,08:00:00,08:15:00,LineaB,LineaB_JMRosas_Oeste_Turn05,Rosas,1,0,0,1
4,2021,1/1/2021,08:00:00,08:15:00,LineaB,LineaB_Malabia_N_Turn02,Malabia,1,0,0,1


In [ ]:
# df_historico_2020.head()

,FECHA,DESDE,HASTA,LINEA,MOLINETE,ESTACION,pax_pagos,pax_pases_pagos,pax_franq,pax_TOTAL
0,01/01/2020,08:00:00,08:15:00,LineaA,LineaA_Acoyte_N_Turn01,Acoyte,1.0,0.0,0.0,1.0
1,01/01/2020,08:00:00,08:15:00,LineaA,LineaA_Carabobo_E_Turn02,Carabobo,6.0,0.0,0.0,6.0
2,01/01/2020,08:00:00,08:15:00,LineaA,LineaA_CBarros_N_Turn03,Castro Barros,3.0,0.0,1.0,4.0
3,01/01/2020,08:00:00,08:15:00,LineaA,LineaA_CBarros_S_Turn02,Castro Barros,2.0,0.0,0.0,2.0
4,01/01/2020,08:00:00,08:15:00,LineaA,LineaA_Congreso_N_Turn03,Congreso,2.0,0.0,0.0,2.0


In [ ]:
# df_historico_2019.head()

,periodo,fecha,desde,hasta,linea,molinete,estacion,pax_pagos,pax_pases_pagos,pax_franq,total
0,201901,2019-01-01,08:00:00,08:15:00,LineaA,LineaA_Lima_N_Turn02,Lima,1.0,0.0,0.0,1.0
1,201901,2019-01-01,08:00:00,08:15:00,LineaA,LineaA_Loria_N_Turn03,Loria,3.0,0.0,0.0,3.0
2,201901,2019-01-01,08:00:00,08:15:00,LineaA,LineaA_Miserere_Q_HALL_Turn01,Plaza Miserere,3.0,0.0,0.0,3.0
3,201901,2019-01-01,08:00:00,08:15:00,LineaA,LineaA_Miserere_S_Turn01,Plaza Miserere,6.0,0.0,0.0,6.0
4,201901,2019-01-01,08:00:00,08:15:00,LineaA,LineaA_Miserere_S_Turn03,Plaza Miserere,10.0,0.0,0.0,10.0


In [ ]:
# df_BaseUnificadaEstaciones.head()

,FECHA,ESTACION,LINEA,DESDE,HORA,CANTIDAD
0,3/1/2020,Acoyte,A,7:45,7,20
1,3/1/2020,Acoyte,A,8:00,8,16
2,3/1/2020,Acoyte,A,8:15,8,22
3,3/1/2020,Acoyte,A,8:30,8,27
4,3/1/2020,Acoyte,A,8:45,8,37


In [ ]:
# for nombre_archivo, df in dataframes_parquet.items():

#     nombre_base = (
#         nombre_archivo
#         .replace('p_', '')
#         .replace('.parquet', '')
#     )

#     nombre_variable = f'df_{nombre_base}'

#     globals()[nombre_variable] = df

#     print(f"Creado: {nombre_variable}")

Creado: df_historico_2018
Creado: df_historico_2021
Creado: df_historico_2020
Creado: df_historico_2019
Creado: df_BaseUnificadaEstaciones
Creado: df_historico_2017


**Análisis exploratorio inicial:**

En primer lugar, se realiza una revisión general de los datasets históricos correspondientes al período 2017–2021. El objetivo es identificar su estructura, dimensiones, tipos de datos, nombres de columnas y posibles registros duplicados antes de realizar cualquier transformación o integración de los datos.

In [ ]:
# historicos = {
#     2017: df_historico_2017,
#     2018: df_historico_2018,
#     2019: df_historico_2019,
#     2020: df_historico_2020,
#     2021: df_historico_2021
# }

# for anio, df in historicos.items():

#     print("=" * 80)
#     print(f"AÑO {anio}")
#     print("=" * 80)

#     print(f"Dimensiones: {df.shape[0]:,} filas x {df.shape[1]} columnas")

#     print("\nColumnas:")
#     print(df.columns.tolist())

#     print("\nTipos de datos:")
#     print(df.dtypes)

#     print("\nDuplicados:")
#     print(f"{df.duplicated().sum():,}")

#     print()

AÑO 2017
Dimensiones: 11,938,476 filas x 12 columnas

Columnas:
['V1', 'FECHA', 'DESDE', 'HASTA', 'LINEA', 'MOLINETE', 'ID_ESTACION', 'ESTACION', 'PAX_PAGOS', 'PAX_PASES_PAGOS', 'PAX_FREQ', 'TOTAL']

Tipos de datos:
V1                  int64
FECHA              object
DESDE              object
HASTA              object
LINEA              object
MOLINETE           object
ID_ESTACION         int64
ESTACION           object
PAX_PAGOS           int64
PAX_PASES_PAGOS     int64
PAX_FREQ            int64
TOTAL               int64
dtype: object

Duplicados:
0

AÑO 2018
Dimensiones: 12,136,454 filas x 11 columnas

Columnas:
['fecha', 'desde', 'hasta', 'linea', 'molinete', 'estacion', 'pax_pagos', 'pax_pases_pagos', 'pax_franq', 'total', 'periodo']

Tipos de datos:
fecha               object
desde               object
hasta               object
linea               object
molinete            object
estacion            object
pax_pagos          float64
pax_pases_pagos    float64
pax_franq          

Los archivos históricos correspondientes al período 2017–2021 presentan una estructura general similar y contienen información de pasajeros por estación, molinete y franja horaria. Sin embargo, se identificaron diferencias entre los años en los nombres de las columnas, uso de mayúsculas y minúsculas, tipos de datos y presencia de campos adicionales. Además, se detectaron registros duplicados en los datasets de 2018 y 2020.

Por lo tanto, antes de unificar los datos será necesario realizar una etapa de homogeneización y limpieza, estandarizando el esquema de los distintos años y analizando los registros duplicados detectados.

**Homogeneización de nombres de columnas:**

Debido a que los archivos históricos presentan diferencias en los nombres de las variables entre los distintos años, se realiza una primera estandarización del esquema para facilitar su posterior análisis e integración.

In [ ]:
# for anio, df in historicos.items():

#     df.columns = (
#         df.columns
#         .str.lower()
#         .str.strip()
#     )

#     print(anio, df.columns.tolist())

2017 ['v1', 'fecha', 'desde', 'hasta', 'linea', 'molinete', 'id_estacion', 'estacion', 'pax_pagos', 'pax_pases_pagos', 'pax_freq', 'total']
2018 ['fecha', 'desde', 'hasta', 'linea', 'molinete', 'estacion', 'pax_pagos', 'pax_pases_pagos', 'pax_franq', 'total', 'periodo']
2019 ['periodo', 'fecha', 'desde', 'hasta', 'linea', 'molinete', 'estacion', 'pax_pagos', 'pax_pases_pagos', 'pax_franq', 'total']
2020 ['fecha', 'desde', 'hasta', 'linea', 'molinete', 'estacion', 'pax_pagos', 'pax_pases_pagos', 'pax_franq', 'pax_total']
2021 ['periodo', 'fecha', 'desde', 'hasta', 'linea', 'molinete', 'estacion', 'pax_pagos', 'pax_pases_pagos', 'pax_franq', 'pax_total']


In [ ]:
# # Homogeneización de nombres de columnas equivalentes

# for anio, df in historicos.items():
#     df.rename(
#         columns={
#             'pax_freq': 'pax_franq',
#             'total': 'pax_total'
#         },
#         inplace=True
#     )

# # Verificar resultado
# for anio, df in historicos.items():
#     print(anio, df.columns.tolist())

2017 ['v1', 'fecha', 'desde', 'hasta', 'linea', 'molinete', 'id_estacion', 'estacion', 'pax_pagos', 'pax_pases_pagos', 'pax_franq', 'pax_total']
2018 ['fecha', 'desde', 'hasta', 'linea', 'molinete', 'estacion', 'pax_pagos', 'pax_pases_pagos', 'pax_franq', 'pax_total', 'periodo']
2019 ['periodo', 'fecha', 'desde', 'hasta', 'linea', 'molinete', 'estacion', 'pax_pagos', 'pax_pases_pagos', 'pax_franq', 'pax_total']
2020 ['fecha', 'desde', 'hasta', 'linea', 'molinete', 'estacion', 'pax_pagos', 'pax_pases_pagos', 'pax_franq', 'pax_total']
2021 ['periodo', 'fecha', 'desde', 'hasta', 'linea', 'molinete', 'estacion', 'pax_pagos', 'pax_pases_pagos', 'pax_franq', 'pax_total']


Homogeneización de nombres: se estandarizaron los nombres de las columnas de los cinco datasets históricos utilizando minúsculas y se unificaron aquellos campos que representan la misma información (pax_freq → pax_franq y total → pax_total). Luego de esta transformación, los datasets presentan un conjunto común de variables, permaneciendo únicamente algunas diferencias particulares asociadas a las columnas v1, id_estacion y periodo.

In [ ]:
# # Análisis de columnas que no están presentes de forma homogénea en todos los años

# for anio, df in historicos.items():

#     print("=" * 60)
#     print(f"AÑO {anio}")

#     if 'periodo' in df.columns:
#         print("\nValores de periodo:")
#         print(
#             df['periodo']
#             .value_counts()
#             .sort_index()
#             .head(20)
#         )

#     if 'v1' in df.columns:
#         print("\nRango de v1:")
#         print(f"Mínimo: {df['v1'].min()}")
#         print(f"Máximo: {df['v1'].max()}")
#         print(f"Valores únicos: {df['v1'].nunique():,}")

#     if 'id_estacion' in df.columns:
#         print("\nID de estaciones:")
#         print(f"Valores únicos: {df['id_estacion'].nunique()}")

#         print(
#             df[['id_estacion', 'estacion']]
#             .drop_duplicates()
#             .sort_values('id_estacion')
#             .head(20)
#         )

AÑO 2017

Rango de v1:
Mínimo: 1
Máximo: 11938476
Valores únicos: 11,938,476

ID de estaciones:
Valores únicos: 86
          id_estacion              estacion
1943090             1               CASEROS
5507168             2                INCLAN
5427348             3            HUMBERTO I
11845322            4             VENEZUELA
7777252             5                  ONCE
0                   6            9 DE JULIO
4238756             7  FACULTAD DE MEDICINA
11340935            8            TRIBUNALES
327126              9                AGUERO
11221854           10      SCALABRINI ORTIZ
8870549            11          PLAZA ITALIA
7945043            12               PALERMO
9912096            13                RETIRO
6336391            14               LAVALLE
3743955            15        DIAGONAL NORTE
627230             16       AVENIDA DE MAYO
7163837            17        MARIANO MORENO
5583695            18         INDEPENDENCIA
2995899            19          CONSTITUCION
50023

**Análisis de columnas no comunes:**

Se analizaron las variables que no se encuentran presentes de forma homogénea en los cinco datasets históricos. La columna v1, disponible únicamente en 2017, corresponde a un índice secuencial de los registros, por lo que no aporta información relevante para el análisis. La variable id_estacion identifica las estaciones mediante 86 valores diferentes, pero solo se encuentra disponible en 2017.

Por otra parte, la variable periodo presenta distintos criterios según el año: en 2018 y 2019 utiliza el formato AAAAMM, en 2021 contiene únicamente el año y no está presente en 2017 ni en 2020. Debido a esta falta de homogeneidad, se optará por generar posteriormente una variable temporal común a partir de la columna fecha.

In [ ]:
# # Análisis del formato original de la columna fecha

# for anio, df in historicos.items():

#     print("=" * 60)
#     print(f"AÑO {anio}")

#     print("Tipo:", df['fecha'].dtype)

#     print("\nEjemplos:")
#     print(
#         df['fecha']
#         .dropna()
#         .astype(str)
#         .sample(
#             min(20, df['fecha'].notna().sum()),
#             random_state=42
#         )
#         .tolist()
#     )

#     print("\nLongitud de los valores:")
#     print(
#         df['fecha']
#         .dropna()
#         .astype(str)
#         .str.len()
#         .value_counts()
#         .sort_index()
#     )

AÑO 2017
Tipo: object

Ejemplos:
['03/11/2017', '28/08/2017', '24/05/2017', '20/08/2017', '05/01/2017', '18/01/2017', '07/12/2017', '26/01/2017', '25/11/2017', '17/11/2017', '29/01/2017', '21/04/2017', '24/02/2017', '29/04/2017', '11/08/2017', '11/06/2017', '31/07/2017', '17/07/2017', '30/08/2017', '04/08/2017']

Longitud de los valores:
fecha
10    11938476
Name: count, dtype: int64
AÑO 2018
Tipo: object

Ejemplos:
['2018-01-09', '2018-11-20', '2018-11-20', '2018-11-27', '2018-10-24', '2018-09-01', '2018-03-19', '2018-04-16', '2018-07-25', '2018-11-27', '2018-08-11', '2018-04-01', '2018-11-25', '2018-06-08', '2018-12-21', '2018-08-16', '2018-04-04', '2018-08-21', '2018-08-13', '2018-06-08']

Longitud de los valores:
fecha
10    12058247
Name: count, dtype: int64
AÑO 2019
Tipo: object

Ejemplos:
['2019-09-18', '2019-12-04', '2019-03-26', '2019-11-22', '2019-08-24', '2019-05-08', '2019-09-30', '2019-08-09', '2019-08-14', '2019-09-24', '2019-05-22', '2019-11-19', '2019-10-19', '2019-02-2

In [ ]:
# # Nulos y cantidad de fechas diferentes antes de realizar transformaciones

# for anio, df in historicos.items():

#     print(
#         f"{anio} | "
#         f"Nulos originales: {df['fecha'].isna().sum():,} | "
#         f"Valores únicos: {df['fecha'].nunique():,}"
#     )

2017 | Nulos originales: 0 | Valores únicos: 365
2018 | Nulos originales: 78,207 | Valores únicos: 365
2019 | Nulos originales: 0 | Valores únicos: 362
2020 | Nulos originales: 1,836 | Valores únicos: 350
2021 | Nulos originales: 0 | Valores únicos: 364


**Análisis inicial de la variable fecha:**

La variable fecha presenta diferencias de formato entre los distintos archivos históricos. En 2017 se utiliza el formato DD/MM/YYYY, mientras que en 2018 y 2019 se observa YYYY-MM-DD. En 2021 las fechas se encuentran expresadas con día, mes y año separados por /, aunque sin una longitud fija debido a la ausencia de ceros a la izquierda en algunos valores.


**Análisis específico de las fechas de 2020:**

Durante la exploración se observó que el dataset correspondiente a 2020 presenta diferentes representaciones de fecha. En particular, algunos registros parecen utilizar el formato DD/MM/YYYY, mientras que otros presentan características compatibles con MM/DD/YYYY.

Debido a que una interpretación incorrecta podría modificar el significado temporal de los registros, se analiza específicamente la posición de los componentes de día y mes antes de realizar cualquier conversión.

In [ ]:
# # Análisis de formatos de fecha en 2020

# fechas_2020 = df_historico_2020['fecha'].dropna().astype(str)

# partes = fechas_2020.str.split('/', expand=True)
# partes.columns = ['parte_1', 'parte_2', 'anio']

# partes['parte_1'] = pd.to_numeric(
#     partes['parte_1'],
#     errors='coerce'
# )

# partes['parte_2'] = pd.to_numeric(
#     partes['parte_2'],
#     errors='coerce'
# )

# # Si la primera parte es mayor a 12, necesariamente representa el día
# dd_mm = partes['parte_1'] > 12

# # Si la segunda parte es mayor a 12, necesariamente representa el día
# mm_dd = partes['parte_2'] > 12

# # Si ambas son <= 12, el formato no puede determinarse solo por el valor
# ambiguas = (
#     (partes['parte_1'] <= 12) &
#     (partes['parte_2'] <= 12)
# )

# print(
#     "Formato DD/MM identificable:",
#     f"{dd_mm.sum():,}"
# )

# print(
#     "Formato MM/DD identificable:",
#     f"{mm_dd.sum():,}"
# )

# print(
#     "Fechas ambiguas:",
#     f"{ambiguas.sum():,}"
# )

Formato DD/MM identificable: 1,723,320
Formato MM/DD identificable: 1,655,090
Fechas ambiguas: 2,400,760


In [ ]:
# # Ejemplos de fechas 2020 conservando el orden original

# fechas_unicas_2020 = (
#     df_historico_2020['fecha']
#     .dropna()
#     .drop_duplicates()
# )

# print("Primeras 30 fechas:")
# print(
#     fechas_unicas_2020
#     .head(30)
#     .tolist()
# )

# print("\nÚltimas 30 fechas:")
# print(
#     fechas_unicas_2020
#     .tail(30)
#     .tolist()
# )

Primeras 30 fechas:
['01/01/2020', '02/01/2020', '03/01/2020', '04/01/2020', '05/01/2020', '06/01/2020', '07/01/2020', '08/01/2020', '09/01/2020', '10/01/2020', '11/01/2020', '12/01/2020', '13/01/2020', '14/01/2020', '15/01/2020', '16/01/2020', '17/01/2020', '18/01/2020', '19/01/2020', '20/01/2020', '21/01/2020', '22/01/2020', '23/01/2020', '24/01/2020', '25/01/2020', '26/01/2020', '27/01/2020', '28/01/2020', '29/01/2020', '30/01/2020']

Últimas 30 fechas:
['24/11/2020', '25/11/2020', '26/11/2020', '27/11/2020', '28/11/2020', '29/11/2020', '30/11/2020', '1/12/2020', '2/12/2020', '11/12/2020', '12/12/2020', '13/12/2020', '14/12/2020', '15/12/2020', '16/12/2020', '17/12/2020', '18/12/2020', '19/12/2020', '20/12/2020', '21/12/2020', '22/12/2020', '23/12/2020', '24/12/2020', '25/12/2020', '26/12/2020', '27/12/2020', '28/12/2020', '29/12/2020', '30/12/2020', '31/12/2020']


In [ ]:
# # Identificar registros de 2020 que son inequívocamente MM/DD/YYYY

# fechas = df_historico_2020['fecha'].dropna().astype(str)

# partes = fechas.str.split('/', expand=True)
# partes.columns = ['p1', 'p2', 'anio']

# partes['p1'] = pd.to_numeric(partes['p1'], errors='coerce')
# partes['p2'] = pd.to_numeric(partes['p2'], errors='coerce')

# # Si la segunda parte es > 12, necesariamente el formato es MM/DD
# indices_mmdd = partes[partes['p2'] > 12].index

# muestra_mmdd = df_historico_2020.loc[
#     indices_mmdd,
#     ['fecha', 'desde', 'hasta', 'linea', 'estacion']
# ]

# print("Primeros registros MM/DD detectados:")
# display(muestra_mmdd.head(20))

# print("\nÚltimos registros MM/DD detectados:")
# display(muestra_mmdd.tail(20))

# print("\nRango de índices:")
# print("Primer índice:", indices_mmdd.min())
# print("Último índice:", indices_mmdd.max())

Primeros registros MM/DD detectados:


,fecha,desde,hasta,linea,estacion
2470080,3/13/2020,05:30:00,5:45:00,LineaA,Acoyte
2470081,3/13/2020,05:30:00,5:45:00,LineaA,Acoyte
2470082,3/13/2020,05:30:00,5:45:00,LineaA,Acoyte
2470083,3/13/2020,05:30:00,5:45:00,LineaA,Acoyte
2470084,3/13/2020,05:30:00,5:45:00,LineaA,Alberti
2470085,3/13/2020,05:30:00,5:45:00,LineaA,Alberti
2470086,3/13/2020,05:30:00,5:45:00,LineaA,Carabobo
2470087,3/13/2020,05:30:00,5:45:00,LineaA,Carabobo
2470088,3/13/2020,05:30:00,5:45:00,LineaA,Carabobo
2470089,3/13/2020,05:30:00,5:45:00,LineaA,Carabobo



Últimos registros MM/DD detectados:


,fecha,desde,hasta,linea,estacion
4960751,10/31/2020,21:45:00,22:00:00,LineaD,Ministro Carranza
4960752,10/31/2020,21:45:00,22:00:00,LineaD,Congreso de Tucuman
4960753,10/31/2020,21:45:00,22:00:00,LineaD,Congreso de Tucuman
4960754,10/31/2020,21:45:00,22:00:00,LineaE,Correo Central
4960755,10/31/2020,21:45:00,22:00:00,LineaE,Correo Central
4960756,10/31/2020,21:45:00,22:00:00,LineaE,Correo Central
4960757,10/31/2020,21:45:00,22:00:00,LineaE,Independencia.H
4960758,10/31/2020,21:45:00,22:00:00,LineaE,Jujuy
4960759,10/31/2020,21:45:00,22:00:00,LineaE,Retiro E
4960760,10/31/2020,21:45:00,22:00:00,LineaH,Facultad de Derecho



Rango de índices:
Primer índice: 2470080
Último índice: 4960770


In [ ]:
# # Distribución de registros inequívocamente MM/DD según la primera parte (mes)

# print(
#     partes.loc[indices_mmdd, 'p1']
#     .value_counts()
#     .sort_index()
# )

p1
3     281699
4     168777
5     192037
6     174840
7     199497
8     198504
9     205448
10    234288
Name: count, dtype: int64


In [ ]:
# # Observar las fechas únicas alrededor del inicio y fin
# # del bloque identificado como MM/DD/YYYY

# for indice in [2470080, 4960770]:

#     print("=" * 70)
#     print(f"Alrededor del índice {indice}")

#     inicio = max(0, indice - 5000)
#     fin = min(len(df_historico_2020), indice + 5000)

#     fechas_entorno = (
#         df_historico_2020.iloc[inicio:fin]['fecha']
#         .dropna()
#         .drop_duplicates()
#         .tolist()
#     )

#     print(fechas_entorno)

Alrededor del índice 2470080
['3/12/2020', '3/13/2020']
Alrededor del índice 4960770
['10/31/2020', '1/11/2020', '2/11/2020']


**Conclusión del análisis de fechas de 2020:**

Se detectó que el histórico correspondiente a 2020 utiliza dos formatos diferentes para representar las fechas. Los registros comprendidos entre el 13 de marzo y el 31 de octubre utilizan el formato MM/DD/YYYY, mientras que los períodos anteriores y posteriores utilizan DD/MM/YYYY.

El análisis del orden de los registros permitió identificar los puntos de transición y resolver también las fechas ambiguas, en las que tanto el día como el mes son menores o iguales a 12. Esta inconsistencia deberá contemplarse durante la normalización para evitar interpretaciones incorrectas de las fechas.

In [ ]:
# # Probar conversión de 2021 sin modificar los datos originales

# fecha_prueba_2021 = pd.to_datetime(
#     df_historico_2021['fecha'],
#     format='%d/%m/%Y',
#     errors='coerce'
# )

# problemas_2021 = fecha_prueba_2021.isna()

# print(
#     "Registros que no pudieron convertirse:",
#     f"{problemas_2021.sum():,}"
# )

# fechas_problematicas_2021 = (
#     df_historico_2021.loc[
#         problemas_2021,
#         'fecha'
#     ]
#     .value_counts()
# )

# print(
#     "Fechas distintas problemáticas:",
#     f"{len(fechas_problematicas_2021):,}"
# )

# print("\nFechas problemáticas:")
# print(fechas_problematicas_2021)

Registros que no pudieron convertirse: 1,043,613
Fechas distintas problemáticas: 59

Fechas problemáticas:
fecha
2021-03-18    21418
2021-03-17    21408
2021-03-23    21265
2021-03-19    21211
2021-03-26    21132
2021-03-29    21130
2021-03-31    21076
2021-03-30    21058
2021-12-03    20942
2021-09-03    20933
2021-08-03    20924
2021-11-03    20881
2021-03-15    20867
2021-03-25    20857
2021-10-03    20841
2021-03-22    20839
2021-03-16    20828
2021-05-03    19876
2021-03-03    19225
2021-02-03    18992
2021-02-25    18973
2021-02-24    18841
2021-02-26    18758
2021-02-23    18752
2021-02-22    18260
2021-02-19    18209
2021-02-18    18141
2021-02-17    18128
2021-12-02    18068
2021-01-03    18001
2021-04-02    17914
2021-09-02    17864
2021-08-02    17850
2021-10-02    17849
2021-11-02    17723
2021-02-02    17712
2021-05-02    17691
2021-01-02    17512
2021-03-27    17430
2021-03-13    17327
2021-03-20    16883
2021-06-03    16334
2021-02-27    15925
2021-02-20    15741
2021-02

In [ ]:
# # Analizar los formatos presentes en 2021

# fechas_2021 = df_historico_2021['fecha'].astype(str)

# formato_barra = fechas_2021.str.contains('/', regex=False)
# formato_guion = fechas_2021.str.contains('-', regex=False)

# print(
#     "Registros con formato DD/MM/YYYY:",
#     f"{formato_barra.sum():,}"
# )

# print(
#     "Registros con formato YYYY-MM-DD:",
#     f"{formato_guion.sum():,}"
# )

# print(
#     "Otros formatos:",
#     f"{(~(formato_barra | formato_guion)).sum():,}"
# )

Registros con formato DD/MM/YYYY: 7,028,067
Registros con formato YYYY-MM-DD: 1,043,613
Otros formatos: 0


In [ ]:
# # Ver los primeros y últimos registros con formato YYYY-MM-DD

# indices_guion = df_historico_2021.index[formato_guion]

# print("Primer índice:", indices_guion.min())
# print("Último índice:", indices_guion.max())

# print("\nPrimeras fechas con guion:")
# print(
#     df_historico_2021.loc[indices_guion, 'fecha']
#     .drop_duplicates()
#     .head(20)
#     .tolist()
# )

# print("\nÚltimas fechas con guion:")
# print(
#     df_historico_2021.loc[indices_guion, 'fecha']
#     .drop_duplicates()
#     .tail(20)
#     .tolist()
# )

Primer índice: 484238
Último índice: 1527850

Primeras fechas con guion:
['2021-01-02', '2021-02-02', '2021-03-02', '2021-04-02', '2021-05-02', '2021-06-02', '2021-07-02', '2021-08-02', '2021-09-02', '2021-10-02', '2021-11-02', '2021-12-02', '2021-02-13', '2021-02-14', '2021-02-15', '2021-02-16', '2021-02-17', '2021-02-18', '2021-02-19', '2021-02-20']

Últimas fechas con guion:
['2021-12-03', '2021-03-13', '2021-03-14', '2021-03-15', '2021-03-16', '2021-03-17', '2021-03-18', '2021-03-19', '2021-03-20', '2021-03-21', '2021-03-22', '2021-03-23', '2021-03-24', '2021-03-25', '2021-03-26', '2021-03-27', '2021-03-28', '2021-03-29', '2021-03-30', '2021-03-31']


**Inconsistencia en las fechas de 2021:**

Se identificaron dos representaciones diferentes de la variable fecha. La mayor parte de los registros utiliza una representación con barras (DD/MM/YYYY), mientras que 1.043.613 registros presentan el formato YYYY-MM-DD. A diferencia de 2020, ambos formatos pueden distinguirse directamente mediante el separador utilizado, por lo que podrán convertirse de forma independiente durante la normalización.

**Conclusión del análisis de fechas de 2021:**

En el histórico de 2021 se identificaron dos representaciones diferentes de la variable fecha. Un total de 7.028.067 registros utiliza una representación con barras (DD/MM/YYYY), mientras que 1.043.613 registros presenta el formato YYYY-MM-DD.

A diferencia de la inconsistencia detectada en 2020, los formatos de 2021 pueden distinguirse directamente mediante el separador utilizado (/ o -). Por lo tanto, durante la normalización ambos grupos podrán convertirse de forma independiente sin necesidad de inferir el orden del día y el mes.

| Año  | Formato                                                 |
| ---- | ------------------------------------------------------- |
| 2017 | `DD/MM/YYYY`                                            |
| 2018 | `YYYY-MM-DD` + 78.207 nulos originales                  |
| 2019 | `YYYY-MM-DD`                                            |
| 2020 | `DD/MM/YYYY` → `MM/DD/YYYY` → `DD/MM/YYYY`              |
| 2021 | `DD/MM/YYYY` y `YYYY-MM-DD` distinguibles por separador |


###**Normalización de la variable fecha:**

A partir de los formatos identificados durante el análisis exploratorio, se procede a convertir la variable fecha a un tipo datetime homogéneo. La conversión se realiza de forma diferenciada según las características detectadas en cada archivo, evitando aplicar un único criterio que pudiera generar interpretaciones incorrectas.

In [ ]:
# # 2017: DD/MM/YYYY
# historicos[2017]['fecha'] = pd.to_datetime(
#     historicos[2017]['fecha'],
#     format='%d/%m/%Y',
#     errors='coerce'
# )

# # 2018 y 2019: YYYY-MM-DD
# for anio in [2018, 2019]:
#     historicos[anio]['fecha'] = pd.to_datetime(
#         historicos[anio]['fecha'],
#         format='%Y-%m-%d',
#         errors='coerce'
#     )

In [ ]:
# # Validación

# for anio in [2017, 2018, 2019]:
#     df = historicos[anio]

#     print(
#         f"{anio} | "
#         f"Tipo: {df['fecha'].dtype} | "
#         f"NaT: {df['fecha'].isna().sum():,} | "
#         f"Fechas únicas: {df['fecha'].nunique():,} | "
#         f"Desde: {df['fecha'].min()} | "
#         f"Hasta: {df['fecha'].max()}"
#     )

2017 | Tipo: datetime64[ns] | NaT: 0 | Fechas únicas: 365 | Desde: 2017-01-01 00:00:00 | Hasta: 2017-12-31 00:00:00
2018 | Tipo: datetime64[ns] | NaT: 78,207 | Fechas únicas: 365 | Desde: 2018-01-01 00:00:00 | Hasta: 2018-12-31 00:00:00
2019 | Tipo: datetime64[ns] | NaT: 0 | Fechas únicas: 362 | Desde: 2019-01-01 00:00:00 | Hasta: 2019-12-31 00:00:00


In [ ]:
# # Normalización de fechas de 2020

# df_2020 = historicos[2020]

# # Bloque identificado con formato MM/DD/YYYY
# bloque_mmdd = (
#     (df_2020.index >= 2470080) &
#     (df_2020.index <= 4960770)
# )

# # Registros fuera del bloque: DD/MM/YYYY
# df_2020.loc[~bloque_mmdd, 'fecha'] = pd.to_datetime(
#     df_2020.loc[~bloque_mmdd, 'fecha'],
#     format='%d/%m/%Y',
#     errors='coerce'
# )

# # Registros dentro del bloque: MM/DD/YYYY
# df_2020.loc[bloque_mmdd, 'fecha'] = pd.to_datetime(
#     df_2020.loc[bloque_mmdd, 'fecha'],
#     format='%m/%d/%Y',
#     errors='coerce'
# )

# # Convertir la columna completa definitivamente a datetime
# df_2020['fecha'] = pd.to_datetime(
#     df_2020['fecha'],
#     errors='coerce'
# )

In [ ]:
# print(
#     f"2020 | "
#     f"Tipo: {df_2020['fecha'].dtype} | "
#     f"NaT: {df_2020['fecha'].isna().sum():,} | "
#     f"Fechas únicas: {df_2020['fecha'].nunique():,} | "
#     f"Desde: {df_2020['fecha'].min()} | "
#     f"Hasta: {df_2020['fecha'].max()}"
# )

2020 | Tipo: datetime64[ns] | NaT: 1,836 | Fechas únicas: 355 | Desde: 2020-01-01 00:00:00 | Hasta: 2020-12-31 00:00:00


In [ ]:
# # Normalización de fechas de 2021

# df_2021 = historicos[2021]

# # Identificar cada formato a partir del separador
# formato_barra = df_2021['fecha'].astype(str).str.contains('/', regex=False)
# formato_guion = df_2021['fecha'].astype(str).str.contains('-', regex=False)

# # DD/MM/YYYY
# df_2021.loc[formato_barra, 'fecha'] = pd.to_datetime(
#     df_2021.loc[formato_barra, 'fecha'],
#     format='%d/%m/%Y',
#     errors='coerce'
# )

# # YYYY-MM-DD
# df_2021.loc[formato_guion, 'fecha'] = pd.to_datetime(
#     df_2021.loc[formato_guion, 'fecha'],
#     format='%Y-%m-%d',
#     errors='coerce'
# )

# # Asegurar un único tipo datetime para toda la columna
# df_2021['fecha'] = pd.to_datetime(
#     df_2021['fecha'],
#     errors='coerce'
# )

In [ ]:
# # Validación
# print(
#     f"2021 | "
#     f"Tipo: {df_2021['fecha'].dtype} | "
#     f"NaT: {df_2021['fecha'].isna().sum():,} | "
#     f"Fechas únicas: {df_2021['fecha'].nunique():,} | "
#     f"Desde: {df_2021['fecha'].min()} | "
#     f"Hasta: {df_2021['fecha'].max()}"
# )

2021 | Tipo: datetime64[ns] | NaT: 0 | Fechas únicas: 344 | Desde: 2021-01-01 00:00:00 | Hasta: 2021-12-31 00:00:00


In [ ]:
# # Recargar 2021 únicamente para comparar las representaciones originales

# ruta_2021 = os.path.join(
#     ruta_carpeta,
#     'p_historico_2021.parquet'
# )

# df_2021_original = pd.read_parquet(ruta_2021)

# df_2021_original.columns = (
#     df_2021_original.columns
#     .str.lower()
#     .str.strip()
# )

# print(
#     "Fechas originales únicas:",
#     df_2021_original['fecha'].nunique()
# )

Fechas originales únicas: 364


In [ ]:
# comparacion_2021 = (
#     df_2021_original[['fecha']]
#     .drop_duplicates()
#     .copy()
# )

# barra = comparacion_2021['fecha'].str.contains('/', regex=False)
# guion = comparacion_2021['fecha'].str.contains('-', regex=False)

# comparacion_2021.loc[barra, 'fecha_normalizada'] = pd.to_datetime(
#     comparacion_2021.loc[barra, 'fecha'],
#     format='%d/%m/%Y'
# )

# comparacion_2021.loc[guion, 'fecha_normalizada'] = pd.to_datetime(
#     comparacion_2021.loc[guion, 'fecha'],
#     format='%Y-%m-%d'
# )

# comparacion_2021['fecha_normalizada'] = pd.to_datetime(
#     comparacion_2021['fecha_normalizada']
# )

In [ ]:
# colisiones_2021 = (
#     comparacion_2021[
#         comparacion_2021.duplicated(
#             subset='fecha_normalizada',
#             keep=False
#         )
#     ]
#     .sort_values('fecha_normalizada')
# )

# print(
#     "Representaciones involucradas:",
#     len(colisiones_2021)
# )

# print(
#     "Fechas normalizadas afectadas:",
#     colisiones_2021['fecha_normalizada'].nunique()
# )

# display(colisiones_2021)

Representaciones involucradas: 40
Fechas normalizadas afectadas: 20


,fecha,fecha_normalizada
8288,2/1/2021,2021-01-02
484238,2021-01-02,2021-01-02
22270,3/1/2021,2021-01-03
942493,2021-01-03,2021-01-03
509845,2021-04-02,2021-04-02
1543809,2/4/2021,2021-04-02
1557669,3/4/2021,2021-04-03
998711,2021-04-03,2021-04-03
519002,2021-05-02,2021-05-02
2078500,2/5/2021,2021-05-02


**Validación de la normalización:**

Luego de convertir ambos formatos se obtuvieron 344 fechas únicas, frente a las 364 representaciones distintas existentes originalmente. Se verificó que esta diferencia no corresponde a una pérdida de información, sino a la existencia de 20 fechas representadas de dos formas diferentes dentro del archivo (por ejemplo, 2/1/2021 y 2021-01-02). Al normalizar ambas representaciones, estas convergen correctamente en una misma fecha. La conversión no generó valores nulos (NaT).

In [ ]:
# # Validación final de la normalización de fechas

# for anio, df in historicos.items():

#     print(
#         f"{anio} | "
#         f"Tipo: {df['fecha'].dtype} | "
#         f"NaT: {df['fecha'].isna().sum():,} | "
#         f"Fechas únicas: {df['fecha'].nunique():,} | "
#         f"Desde: {df['fecha'].min().date()} | "
#         f"Hasta: {df['fecha'].max().date()}"
#     )

2017 | Tipo: datetime64[ns] | NaT: 0 | Fechas únicas: 365 | Desde: 2017-01-01 | Hasta: 2017-12-31
2018 | Tipo: datetime64[ns] | NaT: 78,207 | Fechas únicas: 365 | Desde: 2018-01-01 | Hasta: 2018-12-31
2019 | Tipo: datetime64[ns] | NaT: 0 | Fechas únicas: 362 | Desde: 2019-01-01 | Hasta: 2019-12-31
2020 | Tipo: datetime64[ns] | NaT: 1,836 | Fechas únicas: 355 | Desde: 2020-01-01 | Hasta: 2020-12-31
2021 | Tipo: datetime64[ns] | NaT: 0 | Fechas únicas: 344 | Desde: 2021-01-01 | Hasta: 2021-12-31


**Resultado de la normalización temporal:**

Luego de analizar las particularidades de cada archivo, la variable fecha fue convertida a un formato temporal homogéneo (datetime) para todo el período 2017–2021. El procedimiento contempló explícitamente las diferencias de formato detectadas en 2020 y 2021, evitando la pérdida o interpretación incorrecta de registros.

La validación posterior confirmó que la conversión no introdujo nuevos valores nulos. Los NaT presentes en 2018 y 2020 corresponden a valores que ya se encontraban ausentes en los datos originales. Asimismo, en 2021 se comprobó que la reducción en la cantidad de valores únicos se debe a fechas originalmente representadas mediante más de un formato y no a una pérdida de información.

###**Tratamiento de Duplicados**

In [17]:
# Definir la ruta de 2018: Dataset con Datos Duplicados
ruta_2018 = os.path.join(
    ruta_carpeta,
    "p_historico_2018.parquet"
)

In [22]:
# Obtener una muestra de los grupos duplicados:
consulta_duplicados_2018 = f"""
SELECT
    *,
    COUNT(*) AS cantidad
FROM read_parquet('{ruta_2018}')
GROUP BY ALL
HAVING COUNT(*) > 1
ORDER BY cantidad DESC
LIMIT 20
"""


duckdb.sql(consulta_duplicados_2018).show()


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

┌─────────┬─────────┬─────────┬─────────┬──────────┬──────────┬───────────┬─────────────────┬───────────┬────────┬─────────┬──────────┐
│  fecha  │  desde  │  hasta  │  linea  │ molinete │ estacion │ pax_pagos │ pax_pases_pagos │ pax_franq │ total  │ periodo │ cantidad │
│ varchar │ varchar │ varchar │ varchar │ varchar  │ varchar  │  double   │     double      │  double   │ double │  int64  │  int64   │
├─────────┼─────────┼─────────┼─────────┼──────────┼──────────┼───────────┼─────────────────┼───────────┼────────┼─────────┼──────────┤
│ NULL    │ NULL    │ NULL    │ NULL    │ NULL     │ NULL     │      NULL │            NULL │      NULL │   NULL │  201811 │    49977 │
│ NULL    │ NULL    │ NULL    │ NULL    │ NULL     │ NULL     │      NULL │            NULL │      NULL │   NULL │  201807 │    16491 │
│ NULL    │ NULL    │ NULL    │ NULL    │ NULL     │ NULL     │      NULL │            NULL │      NULL │   NULL │  201809 │    11424 │
│ NULL    │ NULL    │ NULL    │ NULL    │ NULL  

In [27]:
# consulta_nulos_2018 = f"""
# SELECT
#     periodo,
#     COUNT(*) AS cantidad
# FROM read_parquet('{ruta_2018}')
# WHERE fecha IS NULL
# GROUP BY periodo
# ORDER BY periodo
# """

# duckdb.sql(consulta_nulos_2018).show()

┌─────────┬──────────┐
│ periodo │ cantidad │
│  int64  │  int64   │
├─────────┼──────────┤
│  201807 │    16491 │
│  201809 │    11424 │
│  201810 │      315 │
│  201811 │    49977 │
└─────────┴──────────┘



In [ ]:
# consulta_nulos_con_datos_2018 = f"""
# SELECT COUNT(*) AS registros
# FROM read_parquet('{ruta_2018}')
# WHERE fecha IS NULL
#   AND (
#        desde IS NOT NULL
#        OR hasta IS NOT NULL
#        OR linea IS NOT NULL
#        OR molinete IS NOT NULL
#        OR estacion IS NOT NULL
#        OR pax_pagos IS NOT NULL
#        OR pax_pases_pagos IS NOT NULL
#        OR pax_franq IS NOT NULL
#        OR total IS NOT NULL
#   )
# """

# duckdb.sql(consulta_nulos_con_datos_2018).show()

┌───────────┐
│ registros │
│   int64   │
├───────────┤
│         0 │
└───────────┘



**Análisis de registros duplicados y valores nulos de 2018**

El dataset de 2018 presentaba inicialmente 78.203 registros duplicados y 78.207 registros con la variable fecha nula. Al analizar estos casos se comprobó que ambas situaciones se encuentran relacionadas.

Los 78.207 registros con fecha nula corresponden a cuatro grupos asociados a los períodos 201807, 201809, 201810 y 201811. En estos registros, todas las variables descriptivas y de pasajeros (desde, hasta, linea, molinete, estacion, pax_pagos, pax_pases_pagos, pax_franq y total) se encuentran vacías, permaneciendo informado únicamente el campo periodo.

La diferencia entre los 78.207 registros nulos y los 78.203 duplicados se explica porque el método de detección de duplicados conserva la primera ocurrencia de cada uno de los cuatro grupos.

Debido a que estos registros no contienen información sobre estaciones, horarios ni cantidad de pasajeros, se considera que no constituyen observaciones válidas para el análisis y serán excluidos del dataset.

In [29]:
# Eliminación de registros duplicados en DF 2018 y almacenamiento de DF limpio

ruta_2018_limpio = os.path.join(
    ruta_carpeta,
    "p_historico_2018_limpio.parquet"
)

consulta_limpieza_2018 = f"""
COPY (
    SELECT *
    FROM read_parquet('{ruta_2018}')
    WHERE fecha IS NOT NULL
          AND desde IS NOT NULL
          AND hasta IS NOT NULL
          AND linea IS NOT NULL
          AND molinete IS NOT NULL
          AND estacion IS NOT NULL
          AND pax_pagos IS NOT NULL
          AND pax_pases_pagos IS NOT NULL
          AND pax_franq IS NOT NULL
          AND total IS NOT NULL
)
TO '{ruta_2018_limpio}'
(FORMAT PARQUET);
"""

duckdb.sql(consulta_limpieza_2018)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

In [30]:
# Validacion:

consulta_validacion_2018 = f"""
SELECT
    *,
    COUNT(*) AS cantidad
FROM read_parquet('{ruta_2018_limpio}')
GROUP BY ALL
HAVING COUNT(*) > 1
ORDER BY cantidad DESC
LIMIT 20
"""


duckdb.sql(consulta_validacion_2018).show()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

┌─────────┬─────────┬─────────┬─────────┬──────────┬──────────┬───────────┬─────────────────┬───────────┬────────┬─────────┬──────────┐
│  fecha  │  desde  │  hasta  │  linea  │ molinete │ estacion │ pax_pagos │ pax_pases_pagos │ pax_franq │ total  │ periodo │ cantidad │
│ varchar │ varchar │ varchar │ varchar │ varchar  │ varchar  │  double   │     double      │  double   │ double │  int64  │  int64   │
├─────────┴─────────┴─────────┴─────────┴──────────┴──────────┴───────────┴─────────────────┴───────────┴────────┴─────────┴──────────┤
│                                                               0 rows                                                                │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘



In [ ]:
# # Validación

# consulta_validacion_2018 = f"""
# SELECT
#     COUNT(*) AS total_registros,
#     COUNT(*) FILTER (WHERE fecha IS NULL) AS fechas_nulas
# FROM read_parquet('{ruta_2018_limpio}')
# """

# duckdb.sql(consulta_validacion_2018).show()

┌─────────────────┬──────────────┐
│ total_registros │ fechas_nulas │
│      int64      │    int64     │
├─────────────────┼──────────────┤
│        12058247 │            0 │
└─────────────────┴──────────────┘



In [31]:
# Ruta del histórico original de 2020

ruta_2020 = os.path.join(
    ruta_carpeta,
    "p_historico_2020.parquet"
)

In [32]:
# Muestra de registros duplicados de 2020

consulta_duplicados_2020 = f"""
SELECT
    *,
    COUNT(*) AS cantidad
FROM read_parquet('{ruta_2020}')
GROUP BY ALL
HAVING COUNT(*) > 1
ORDER BY cantidad DESC
LIMIT 20
"""

duckdb.sql(consulta_duplicados_2020).show()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

┌───────────┬──────────┬──────────┬─────────┬───────────────────────────────┬─────────────────────┬───────────┬─────────────────┬───────────┬───────────┬──────────┐
│   FECHA   │  DESDE   │  HASTA   │  LINEA  │           MOLINETE            │      ESTACION       │ pax_pagos │ pax_pases_pagos │ pax_franq │ pax_TOTAL │ cantidad │
│  varchar  │ varchar  │ varchar  │ varchar │            varchar            │       varchar       │  double   │     double      │  double   │  double   │  int64   │
├───────────┼──────────┼──────────┼─────────┼───────────────────────────────┼─────────────────────┼───────────┼─────────────────┼───────────┼───────────┼──────────┤
│ NULL      │ NULL     │ NULL     │ NULL    │ NULL                          │ NULL                │      NULL │            NULL │      NULL │      NULL │     1836 │
│ 6/11/2020 │ 09:45:00 │ 10:00:00 │ LineaA  │ LineaA_Congreso_N_Turn01      │ Congreso            │       1.0 │             0.0 │       0.0 │       1.0 │        2 │
│ 6/11/202

In [ ]:
# consulta_fechas_duplicadas_2020 = f"""
# WITH grupos_duplicados AS (
#     SELECT
#         FECHA,
#         DESDE,
#         HASTA,
#         LINEA,
#         MOLINETE,
#         ESTACION,
#         pax_pagos,
#         pax_pases_pagos,
#         pax_franq,
#         pax_TOTAL,
#         COUNT(*) AS cantidad
#     FROM read_parquet('{ruta_2020}')
#     WHERE FECHA IS NOT NULL
#     GROUP BY ALL
#     HAVING COUNT(*) > 1
# )

# SELECT
#     FECHA,
#     COUNT(*) AS grupos_duplicados,
#     SUM(cantidad - 1) AS registros_duplicados_excedentes
# FROM grupos_duplicados
# GROUP BY FECHA
# ORDER BY registros_duplicados_excedentes DESC
# """

# duplicados_por_fecha_2020 = duckdb.sql(
#     consulta_fechas_duplicadas_2020
# ).df()

# display(duplicados_por_fecha_2020)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,FECHA,grupos_duplicados,registros_duplicados_excedentes
0,8/12/2020,880,880.0
1,9/11/2020,848,848.0
2,6/12/2020,783,783.0
3,8/11/2020,759,759.0
4,5/12/2020,683,683.0
5,6/11/2020,650,650.0
6,5/11/2020,621,621.0
7,7/11/2020,504,504.0
8,10/11/2020,393,393.0
9,7/12/2020,376,376.0


In [ ]:
# consulta_resumen_duplicados_2020 = f"""
# WITH grupos_duplicados AS (
#     SELECT
#         COUNT(*) AS cantidad
#     FROM read_parquet('{ruta_2020}')
#     WHERE FECHA IS NOT NULL
#     GROUP BY
#         FECHA,
#         DESDE,
#         HASTA,
#         LINEA,
#         MOLINETE,
#         ESTACION,
#         pax_pagos,
#         pax_pases_pagos,
#         pax_franq,
#         pax_TOTAL
#     HAVING COUNT(*) > 1
# )

# SELECT
#     COUNT(*) AS grupos_duplicados,
#     SUM(cantidad) AS filas_involucradas,
#     SUM(cantidad - 1) AS duplicados_excedentes,
#     MAX(cantidad) AS max_repeticiones
# FROM grupos_duplicados
# """

# duckdb.sql(
#     consulta_resumen_duplicados_2020
# ).show()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

┌───────────────────┬────────────────────┬───────────────────────┬──────────────────┐
│ grupos_duplicados │ filas_involucradas │ duplicados_excedentes │ max_repeticiones │
│       int64       │       int128       │        int128         │      int64       │
├───────────────────┼────────────────────┼───────────────────────┼──────────────────┤
│              7947 │              15894 │                  7947 │                2 │
└───────────────────┴────────────────────┴───────────────────────┴──────────────────┘



**Análisis de registros duplicados y valores nulos de 2020:**

El histórico de 2020 presentaba inicialmente 9.782 registros duplicados. El análisis permitió identificar dos situaciones diferentes.

Por un lado, se encontraron 1.836 registros completamente vacíos, incluyendo la variable fecha. Estas filas conforman un único grupo de registros idénticos y explican 1.835 de los duplicados detectados.

Por otro lado, se identificaron 7.947 grupos de observaciones duplicadas con información válida, correspondientes a 15.894 filas involucradas. Cada observación aparece exactamente dos veces y los casos se concentran entre noviembre y diciembre de 2020. Los registros coinciden en fecha, intervalo horario, línea, molinete, estación y cantidad de pasajeros.

Debido a que mantener ambas ocurrencias produciría una doble contabilización de pasajeros, se conservará una única observación de cada grupo duplicado. Asimismo, los 1.836 registros completamente vacíos serán excluidos por no contener información útil para el análisis.

In [36]:
# Ruta de los archivos procesados

ruta_limpios = os.path.join(
    ruta_carpeta,
    "datos_limpios"
)

os.makedirs(ruta_limpios, exist_ok=True)

print(ruta_limpios)

/content/drive/MyDrive/Datos_csv/datos_limpios


In [38]:
# Generación de DF 2020 limpio

ruta_2020_limpio = os.path.join(
    ruta_limpios,
    "p_historico_2020_limpio.parquet"
)

consulta_limpieza_2020 = f"""
COPY (
    SELECT DISTINCT * FROM read_parquet('{ruta_2020}')
    WHERE FECHA IS NOT NULL
          AND DESDE IS NOT NULL
          AND HASTA IS NOT NULL
          AND LINEA IS NOT NULL
          AND MOLINETE IS NOT NULL
          AND ESTACION IS NOT NULL
          AND pax_pagos IS NOT NULL
          AND pax_pases_pagos IS NOT NULL
          AND pax_franq IS NOT NULL
          AND pax_TOTAL IS NOT NULL
)
TO '{ruta_2020_limpio}'
(FORMAT PARQUET);
"""

duckdb.sql(consulta_limpieza_2020)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

In [39]:
# Validacion

consulta_duplicados_2020 = f"""
SELECT
    *,
    COUNT(*) AS cantidad
FROM read_parquet('{ruta_2020_limpio}')
GROUP BY ALL
HAVING COUNT(*) > 1
ORDER BY cantidad DESC
LIMIT 20
"""

duckdb.sql(consulta_duplicados_2020).show()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

┌─────────┬─────────┬─────────┬─────────┬──────────┬──────────┬───────────┬─────────────────┬───────────┬───────────┬──────────┐
│  FECHA  │  DESDE  │  HASTA  │  LINEA  │ MOLINETE │ ESTACION │ pax_pagos │ pax_pases_pagos │ pax_franq │ pax_TOTAL │ cantidad │
│ varchar │ varchar │ varchar │ varchar │ varchar  │ varchar  │  double   │     double      │  double   │  double   │  int64   │
├─────────┴─────────┴─────────┴─────────┴──────────┴──────────┴───────────┴─────────────────┴───────────┴───────────┴──────────┤
│                                                            0 rows                                                            │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘



In [ ]:
# consulta_validacion_2020 = f"""
# SELECT
#     COUNT(*) AS total_registros,
#     COUNT(*) FILTER (WHERE FECHA IS NULL) AS fechas_nulas
# FROM read_parquet('{ruta_2020_limpio}')
# """

# duckdb.sql(consulta_validacion_2020).show()

┌─────────────────┬──────────────┐
│ total_registros │ fechas_nulas │
│      int64      │    int64     │
├─────────────────┼──────────────┤
│         5771223 │            0 │
└─────────────────┴──────────────┘



###**Normalizacion de DF**

####**2017**

In [40]:
ruta_2017 = os.path.join(
    ruta_carpeta,
    "p_historico_2017.parquet"
)

ruta_2017_final = os.path.join(
    ruta_limpios,
    "p_historico_2017_final.parquet"
)

In [41]:
consulta_final_2017 = f"""
COPY (
    SELECT
        strptime(FECHA, '%d/%m/%Y')::DATE AS fecha,
        DESDE AS desde,
        HASTA AS hasta,
        LINEA AS linea,
        MOLINETE AS molinete,
        ESTACION AS estacion,
        PAX_PAGOS::DOUBLE AS pax_pagos,
        PAX_PASES_PAGOS::DOUBLE AS pax_pases_pagos,
        PAX_FREQ::DOUBLE AS pax_franq,
        TOTAL::DOUBLE AS pax_total
    FROM read_parquet('{ruta_2017}')
)
TO '{ruta_2017_final}'
(FORMAT PARQUET);
"""

duckdb.sql(consulta_final_2017)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

In [42]:
consulta_validacion_2017 = f"""
SELECT
    COUNT(*) AS registros,
    COUNT(*) FILTER (WHERE fecha IS NULL) AS fechas_nulas,
    COUNT(DISTINCT fecha) AS fechas_unicas,
    MIN(fecha) AS fecha_min,
    MAX(fecha) AS fecha_max
FROM read_parquet('{ruta_2017_final}')
"""

duckdb.sql(consulta_validacion_2017).show()

┌───────────┬──────────────┬───────────────┬────────────┬────────────┐
│ registros │ fechas_nulas │ fechas_unicas │ fecha_min  │ fecha_max  │
│   int64   │    int64     │     int64     │    date    │    date    │
├───────────┼──────────────┼───────────────┼────────────┼────────────┤
│  11938476 │            0 │           365 │ 2017-01-01 │ 2017-12-31 │
└───────────┴──────────────┴───────────────┴────────────┴────────────┘



####**2018**

In [43]:
ruta_2018_final = os.path.join(
    ruta_limpios,
    "p_historico_2018_final.parquet"
)

consulta_final_2018 = f"""
COPY (
    SELECT
        strptime(fecha, '%Y-%m-%d')::DATE AS fecha,
        desde,
        hasta,
        linea,
        molinete,
        estacion,
        pax_pagos::DOUBLE AS pax_pagos,
        pax_pases_pagos::DOUBLE AS pax_pases_pagos,
        pax_franq::DOUBLE AS pax_franq,
        total::DOUBLE AS pax_total
    FROM read_parquet('{ruta_2018_limpio}')
)
TO '{ruta_2018_final}'
(FORMAT PARQUET);
"""

duckdb.sql(consulta_final_2018)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

In [44]:
consulta_validacion_2018_final = f"""
SELECT
    COUNT(*) AS registros,
    COUNT(*) FILTER (WHERE fecha IS NULL) AS fechas_nulas,
    COUNT(DISTINCT fecha) AS fechas_unicas,
    MIN(fecha) AS fecha_min,
    MAX(fecha) AS fecha_max
FROM read_parquet('{ruta_2018_final}')
"""

duckdb.sql(consulta_validacion_2018_final).show()

┌───────────┬──────────────┬───────────────┬────────────┬────────────┐
│ registros │ fechas_nulas │ fechas_unicas │ fecha_min  │ fecha_max  │
│   int64   │    int64     │     int64     │    date    │    date    │
├───────────┼──────────────┼───────────────┼────────────┼────────────┤
│  12058191 │            0 │           365 │ 2018-01-01 │ 2018-12-31 │
└───────────┴──────────────┴───────────────┴────────────┴────────────┘



####**2019**

In [45]:
ruta_2019 = os.path.join(
    ruta_carpeta,
    "p_historico_2019.parquet"
)

ruta_2019_final = os.path.join(
    ruta_limpios,
    "p_historico_2019_final.parquet"
)

consulta_final_2019 = f"""
COPY (
    SELECT
        strptime(fecha, '%Y-%m-%d')::DATE AS fecha,
        desde,
        hasta,
        linea,
        molinete,
        estacion,
        pax_pagos::DOUBLE AS pax_pagos,
        pax_pases_pagos::DOUBLE AS pax_pases_pagos,
        pax_franq::DOUBLE AS pax_franq,
        total::DOUBLE AS pax_total
    FROM read_parquet('{ruta_2019}')
)
TO '{ruta_2019_final}'
(FORMAT PARQUET);
"""

duckdb.sql(consulta_final_2019)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

In [50]:
consulta_validacion_2019 = f"""
SELECT
    COUNT(*) AS registros,
    COUNT(*) FILTER (WHERE fecha IS NULL) AS fechas_nulas,
    COUNT(DISTINCT fecha) AS fechas_unicas,
    MIN(fecha) AS fecha_min,
    MAX(fecha) AS fecha_max
FROM read_parquet('{ruta_2019_final}')
"""

duckdb.sql(consulta_validacion_2019).show()

┌───────────┬──────────────┬───────────────┬────────────┬────────────┐
│ registros │ fechas_nulas │ fechas_unicas │ fecha_min  │ fecha_max  │
│   int64   │    int64     │     int64     │    date    │    date    │
├───────────┼──────────────┼───────────────┼────────────┼────────────┤
│  12662343 │            0 │           362 │ 2019-01-01 │ 2019-12-31 │
└───────────┴──────────────┴───────────────┴────────────┴────────────┘



####**2020**

In [53]:
ruta_2020_final = os.path.join(
    ruta_limpios,
    "p_historico_2020_final.parquet"
)

consulta_final_2020 = f"""
COPY (
    WITH original AS (
        SELECT
            *,
            row_number() OVER () - 1 AS indice_original
        FROM read_parquet('{ruta_2020_limpio}')
    ),
    normalizado AS (
        SELECT
            COALESCE(
                TRY_STRPTIME(FECHA, '%m/%d/%Y'), -- Try MM/DD/YYYY first
                TRY_STRPTIME(FECHA, '%d/%m/%Y')  -- Then try DD/MM/YYYY
            )::DATE AS fecha,
            DESDE AS desde,
            HASTA AS hasta,
            LINEA AS linea,
            MOLINETE AS molinete,
            ESTACION AS estacion,
            pax_pagos::DOUBLE AS pax_pagos,
            pax_pases_pagos::DOUBLE AS pax_pases_pagos,
            pax_franq::DOUBLE AS pax_franq,
            pax_TOTAL::DOUBLE AS pax_total
        FROM original
        WHERE FECHA IS NOT NULL
    )
    SELECT DISTINCT *
    FROM normalizado
)
TO '{ruta_2020_final}'
(FORMAT PARQUET);
"""

duckdb.sql(consulta_final_2020)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

In [56]:
consulta_validacion_2020_final = f"""
SELECT
    COUNT(*) AS registros,
    COUNT(*) FILTER (WHERE fecha IS NULL) AS fechas_nulas,
    COUNT(DISTINCT fecha) AS fechas_unicas,
    MIN(fecha) AS fecha_min,
    MAX(fecha) AS fecha_max
FROM read_parquet('{ruta_2020_final}')
"""

duckdb.sql(consulta_validacion_2020_final).show()

┌───────────┬──────────────┬───────────────┬────────────┬────────────┐
│ registros │ fechas_nulas │ fechas_unicas │ fecha_min  │ fecha_max  │
│   int64   │    int64     │     int64     │    date    │    date    │
├───────────┼──────────────┼───────────────┼────────────┼────────────┤
│   5768076 │            0 │           334 │ 2020-01-01 │ 2020-12-31 │
└───────────┴──────────────┴───────────────┴────────────┴────────────┘



In [58]:
# consulta_orden_2020 = f"""
# WITH datos AS (
#     SELECT
#         row_number() OVER () - 1 AS indice,
#         FECHA
#     FROM read_parquet('{ruta_2020_limpio}')
# )
# SELECT indice, FECHA
# FROM datos
# WHERE indice BETWEEN 2465000 AND 2475000
#   AND (
#       FECHA IN ('3/12/2020', '3/13/2020')
#       OR FECHA IN ('12/3/2020', '13/3/2020')
#   )
# ORDER BY indice
# LIMIT 20
# """

# duckdb.sql(consulta_orden_2020).show()

┌────────┬───────┐
│ indice │ fecha │
│ int64  │ date  │
├────────┴───────┤
│     0 rows     │
└────────────────┘



In [ ]:
# if os.path.exists(ruta_2020_final):
#     os.remove(ruta_2020_final)

In [ ]:
# consulta_final_2020 = f"""
# COPY (
#     WITH original AS (
#         SELECT
#             row_number() OVER () - 1 AS indice_original,
#             *
#         FROM read_parquet('{ruta_2020}')
#     ),

#     deduplicado AS (
#         SELECT
#             MIN(indice_original) AS indice_original,
#             FECHA,
#             DESDE,
#             HASTA,
#             LINEA,
#             MOLINETE,
#             ESTACION,
#             pax_pagos,
#             pax_pases_pagos,
#             pax_franq,
#             pax_TOTAL
#         FROM original
#         WHERE FECHA IS NOT NULL
#         GROUP BY
#             FECHA,
#             DESDE,
#             HASTA,
#             LINEA,
#             MOLINETE,
#             ESTACION,
#             pax_pagos,
#             pax_pases_pagos,
#             pax_franq,
#             pax_TOTAL
#     )

#     SELECT
#         CASE
#             WHEN indice_original BETWEEN 2470080 AND 4960770
#                 THEN strptime(FECHA, '%m/%d/%Y')::DATE
#             ELSE strptime(FECHA, '%d/%m/%Y')::DATE
#         END AS fecha,
#         DESDE AS desde,
#         HASTA AS hasta,
#         LINEA AS linea,
#         MOLINETE AS molinete,
#         ESTACION AS estacion,
#         pax_pagos::DOUBLE AS pax_pagos,
#         pax_pases_pagos::DOUBLE AS pax_pases_pagos,
#         pax_franq::DOUBLE AS pax_franq,
#         pax_TOTAL::DOUBLE AS pax_total
#     FROM deduplicado
# )
# TO '{ruta_2020_final}'
# (FORMAT PARQUET);
# """

# duckdb.sql(consulta_final_2020)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

In [ ]:
# duckdb.sql(consulta_validacion_2020_final).show()

┌───────────┬──────────────┬───────────────┬────────────┬────────────┐
│ registros │ fechas_nulas │ fechas_unicas │ fecha_min  │ fecha_max  │
│   int64   │    int64     │     int64     │    date    │    date    │
├───────────┼──────────────┼───────────────┼────────────┼────────────┤
│   5771223 │            0 │           355 │ 2020-01-01 │ 2020-12-31 │
└───────────┴──────────────┴───────────────┴────────────┴────────────┘



####**2021**

In [59]:
ruta_2021 = os.path.join(
    ruta_carpeta,
    "p_historico_2021.parquet"
)

ruta_2021_final = os.path.join(
    ruta_limpios,
    "p_historico_2021_final.parquet"
)

In [60]:
consulta_final_2021 = f"""
COPY (
    SELECT
        CASE
            WHEN FECHA LIKE '%/%'
                THEN strptime(FECHA, '%d/%m/%Y')::DATE
            WHEN FECHA LIKE '%-%'
                THEN strptime(FECHA, '%Y-%m-%d')::DATE
        END AS fecha,
        DESDE AS desde,
        HASTA AS hasta,
        LINEA AS linea,
        MOLINETE AS molinete,
        ESTACION AS estacion,
        pax_pagos::DOUBLE AS pax_pagos,
        pax_pases_pagos::DOUBLE AS pax_pases_pagos,
        pax_franq::DOUBLE AS pax_franq,
        pax_TOTAL::DOUBLE AS pax_total
    FROM read_parquet('{ruta_2021}')
)
TO '{ruta_2021_final}'
(FORMAT PARQUET);
"""

duckdb.sql(consulta_final_2021)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

In [61]:
consulta_validacion_2021 = f"""
SELECT
    COUNT(*) AS registros,
    COUNT(*) FILTER (WHERE fecha IS NULL) AS fechas_nulas,
    COUNT(DISTINCT fecha) AS fechas_unicas,
    MIN(fecha) AS fecha_min,
    MAX(fecha) AS fecha_max
FROM read_parquet('{ruta_2021_final}')
"""

duckdb.sql(consulta_validacion_2021).show()

┌───────────┬──────────────┬───────────────┬────────────┬────────────┐
│ registros │ fechas_nulas │ fechas_unicas │ fecha_min  │ fecha_max  │
│   int64   │    int64     │     int64     │    date    │    date    │
├───────────┼──────────────┼───────────────┼────────────┼────────────┤
│   8071680 │            0 │           344 │ 2021-01-01 │ 2021-12-31 │
└───────────┴──────────────┴───────────────┴────────────┴────────────┘



###**Control de versión final**

In [62]:
rutas_finales = {
    2017: ruta_2017_final,
    2018: ruta_2018_final,
    2019: ruta_2019_final,
    2020: ruta_2020_final,
    2021: ruta_2021_final
}

for anio, ruta in rutas_finales.items():

    print("=" * 60)
    print(f"AÑO {anio}")

    consulta = f"""
    DESCRIBE
    SELECT *
    FROM read_parquet('{ruta}')
    """

    duckdb.sql(consulta).show()

AÑO 2017
┌─────────────────┬─────────────┬─────────┬─────────┬─────────┬─────────┐
│   column_name   │ column_type │  null   │   key   │ default │  extra  │
│     varchar     │   varchar   │ varchar │ varchar │ varchar │ varchar │
├─────────────────┼─────────────┼─────────┼─────────┼─────────┼─────────┤
│ fecha           │ DATE        │ YES     │ NULL    │ NULL    │ NULL    │
│ desde           │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ hasta           │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ linea           │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ molinete        │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ estacion        │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ pax_pagos       │ DOUBLE      │ YES     │ NULL    │ NULL    │ NULL    │
│ pax_pases_pagos │ DOUBLE      │ YES     │ NULL    │ NULL    │ NULL    │
│ pax_franq       │ DOUBLE      │ YES     │ NULL    │ NULL    │ NULL    │
│ pax_total       │ DOUBLE   

In [63]:
for anio, ruta in rutas_finales.items():

    consulta = f"""
    SELECT COUNT(*) AS registros
    FROM read_parquet('{ruta}')
    """

    cantidad = duckdb.sql(consulta).fetchone()[0]

    print(f"{anio}: {cantidad:,} registros")

2017: 11,938,476 registros
2018: 12,058,191 registros
2019: 12,662,343 registros
2020: 5,768,076 registros
2021: 8,071,680 registros


###**Unificación de los datasets**

In [64]:
ruta_historico_unificado = os.path.join(
    ruta_limpios,
    "historico_2017_2021_unificado.parquet"
)

consulta_unificacion = f"""
COPY (
    SELECT * FROM read_parquet('{ruta_2017_final}')
    UNION ALL
    SELECT * FROM read_parquet('{ruta_2018_final}')
    UNION ALL
    SELECT * FROM read_parquet('{ruta_2019_final}')
    UNION ALL
    SELECT * FROM read_parquet('{ruta_2020_final}')
    UNION ALL
    SELECT * FROM read_parquet('{ruta_2021_final}')
)
TO '{ruta_historico_unificado}'
(FORMAT PARQUET);
"""

duckdb.sql(consulta_unificacion)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

In [65]:
consulta_validacion_unificado = f"""
SELECT
    COUNT(*) AS registros,
    COUNT(*) FILTER (WHERE fecha IS NULL) AS fechas_nulas,
    COUNT(DISTINCT fecha) AS fechas_unicas,
    MIN(fecha) AS fecha_min,
    MAX(fecha) AS fecha_max
FROM read_parquet('{ruta_historico_unificado}')
"""

duckdb.sql(consulta_validacion_unificado).show()

┌───────────┬──────────────┬───────────────┬────────────┬────────────┐
│ registros │ fechas_nulas │ fechas_unicas │ fecha_min  │ fecha_max  │
│   int64   │    int64     │     int64     │    date    │    date    │
├───────────┼──────────────┼───────────────┼────────────┼────────────┤
│  50498766 │            0 │          1770 │ 2017-01-01 │ 2021-12-31 │
└───────────┴──────────────┴───────────────┴────────────┴────────────┘



In [66]:
consulta_por_anio = f"""
SELECT
    YEAR(fecha) AS anio,
    COUNT(*) AS registros,
    COUNT(DISTINCT fecha) AS fechas_unicas,
    MIN(fecha) AS fecha_min,
    MAX(fecha) AS fecha_max
FROM read_parquet('{ruta_historico_unificado}')
GROUP BY YEAR(fecha)
ORDER BY anio
"""

duckdb.sql(consulta_por_anio).show()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

┌───────┬───────────┬───────────────┬────────────┬────────────┐
│ anio  │ registros │ fechas_unicas │ fecha_min  │ fecha_max  │
│ int64 │   int64   │     int64     │    date    │    date    │
├───────┼───────────┼───────────────┼────────────┼────────────┤
│  2017 │  11938476 │           365 │ 2017-01-01 │ 2017-12-31 │
│  2018 │  12058191 │           365 │ 2018-01-01 │ 2018-12-31 │
│  2019 │  12662343 │           362 │ 2019-01-01 │ 2019-12-31 │
│  2020 │   5768076 │           334 │ 2020-01-01 │ 2020-12-31 │
│  2021 │   8071680 │           344 │ 2021-01-01 │ 2021-12-31 │
└───────┴───────────┴───────────────┴────────────┴────────────┘



**Resultado de la limpieza y unificación de los históricos:**

Luego del análisis exploratorio se homogeneizaron los cinco datasets correspondientes al período 2017–2021. Se estandarizaron los nombres y tipos de las variables, se normalizó la representación de las fechas y se trataron los registros nulos y duplicados identificados en 2018 y 2020.

Como resultado, los cinco históricos presentan un esquema común compuesto por las variables de fecha, intervalo horario, línea, molinete, estación y cantidad de pasajeros. Posteriormente, los archivos fueron integrados mediante una operación UNION ALL, obteniendo un dataset consolidado de 50.501.969 registros, sin valores nulos en la variable fecha y con información comprendida entre enero de 2017 y diciembre de 2021.

Debido al volumen de los datos, el procesamiento se realizó directamente sobre archivos Parquet mediante DuckDB, evitando cargar simultáneamente la totalidad de los registros en memoria.

In [ ]:
## Hasta ACA LLEGUE - TODO Validado - JAVI & LEO

In [ ]:
# Normalizacion de columnas para el dataSet 2017

# 1. Renombrar la columna V1 a 'periodo'
# df_historico_2017 = df_historico_2017.rename(columns={'V1': 'periodo'})
# print("Columna 'V1' renombrada a 'periodo'.")

# 2. Extraer el año de la columna FECHA
# Primero, convertir la columna 'FECHA' a formato datetime si no lo está
# Intentamos varios formatos comunes en caso de inconsistencia
# try:
#     df_historico_2017['FECHA'] = pd.to_datetime(df_historico_2017['FECHA'], format='%d/%m/%Y', errors='coerce')
# except ValueError:
    # Si falla el formato %d/%m/%Y, intentamos con %Y-%m-%d o %m/%d/%Y
#     df_historico_2017['FECHA'] = pd.to_datetime(df_historico_2017['FECHA'], errors='coerce')

# Extraer el año y guardar en una nueva columna o sobrescribir 'periodo'
# Asumiendo que 'periodo' ahora se refiere al año, podemos actualizarla o crear una nueva
# Si 'periodo' ya existe por el renombramiento, la actualizamos con el año
# Si el usuario quiere una nueva columna, se podría llamar 'año' o similar
# Por la descripción "extraigas el año de la columna fecha", lo más lógico es una nueva columna 'año' o actualizar 'periodo'
# Voy a actualizar 'periodo' ya que el nombre 'V1' que cambió a 'periodo' no es significativo y el año es un periodo.
# Si 'FECHA' no pudo ser convertida, resultará en NaT. Nos aseguramos de manejarlo.
# df_historico_2017['periodo'] = df_historico_2017['FECHA'].dt.year
# print("Año extraído de la columna 'FECHA' y guardado en 'periodo'.")

# 3. Eliminar la columna ID_ESTACION
# if 'ID_ESTACION' in df_historico_2017.columns:
#     df_historico_2017 = df_historico_2017.drop(columns=['ID_ESTACION'])
#     print("Columna 'ID_ESTACION' eliminada.")
# else:
#     print("La columna 'ID_ESTACION' no se encontró en df_historico_2017.")

# print("\nPrimeras filas del DataFrame df_historico_2017 modificado:")
# print(df_historico_2017.head())

# print("\nInformación del DataFrame df_historico_2017 modificado:")
# df_historico_2017.info()


Modificando df_historico_2017...
Columna 'V1' renombrada a 'periodo'.
Año extraído de la columna 'FECHA' y guardado en 'periodo'.
La columna 'ID_ESTACION' no se encontró en df_historico_2017.

Primeras filas del DataFrame df_historico_2017 modificado:
   periodo      FECHA     DESDE     HASTA LINEA                 MOLINETE  \
0     2017 2017-01-01  08:00:00  08:15:00     D  LINEA_D_9JULIO_S_TURN02   
1     2017 2017-01-01  08:00:00  08:15:00     D  LINEA_D_9JULIO_N_TURN02   
2     2017 2017-01-01  08:00:00  08:15:00     D  LINEA_D_9JULIO_S_TURN01   
3     2017 2017-01-01  08:15:00  08:30:00     D  LINEA_D_9JULIO_S_TURN02   
4     2017 2017-01-01  08:15:00  08:30:00     D  LINEA_D_9JULIO_S_TURN01   

     ESTACION  PAX_PAGOS  PAX_PASES_PAGOS  PAX_FREQ  TOTAL  
0  9 DE JULIO          1                0         0      1  
1  9 DE JULIO          1                0         0      1  
2  9 DE JULIO          1                0         0      1  
3  9 DE JULIO          1                0      

In [ ]:
# Normalizacion del DataSet 2018

# print("Modificando df_historico_2018...")

# Extraer el año de la columna 'fecha' sin convertir la columna a datetime.
# Asumiendo que 'fecha' tiene un formato donde el año está al principio (ej: 'YYYY-MM-DD').
# Convertimos la columna a string para asegurar el uso de métodos de string y extraemos los primeros 4 caracteres.
# Esto crea la columna 'Periodo' como tipo 'object' (string).
# df_historico_2018['Periodo'] = df_historico_2018['fecha'].astype(str).str[:4]
# print("Columna 'Periodo' creada y año extraído de 'fecha' sin modificar el tipo de 'fecha'.")

# Reordenar las columnas para colocar 'Periodo' al principio
# Obtener todas las columnas actuales
# columnas = df_historico_2018.columns.tolist()

# Asegurarse de que 'Periodo' no esté duplicado en la lista antes de insertarlo,
# en caso de que la celda se ejecute múltiples veces.
# if 'Periodo' in columnas:
#     columnas.remove('Periodo')

# Insertar 'Periodo' al inicio de la lista de columnas
# columnas.insert(0, 'Periodo')

# Reindexar el DataFrame con el nuevo orden de columnas
# df_historico_2018 = df_historico_2018[columnas]
# print("Columna 'Periodo' movida al inicio del DataFrame.")

# print("\nPrimeras filas del DataFrame df_historico_2018 modificado:")
# print(df_historico_2018.head())

# print("\nInformación del DataFrame df_historico_2018 modificado:")
# df_historico_2018.info()


Modificando df_historico_2018...
Columna 'Periodo' creada y año extraído de 'fecha' sin modificar el tipo de 'fecha'.
Columna 'Periodo' movida al inicio del DataFrame.

Primeras filas del DataFrame df_historico_2018 modificado:
  Periodo      fecha     desde     hasta   linea                 molinete  \
0    2018 2018-01-01  08:00:00  08:15:00  LineaA  LineaA_CBarros_S_Turn01   
1    2018 2018-01-01  08:00:00  08:15:00  LineaA     LineaA_Lima_S_Turn03   
2    2018 2018-01-01  08:00:00  08:15:00  LineaA      LineaA_Pasco_Turn01   
3    2018 2018-01-01  08:00:00  08:15:00  LineaA     LineaA_Peru_S_Turn01   
4    2018 2018-01-01  08:00:00  08:15:00  LineaA   LineaA_PJunta_S_Turn02   

        estacion  pax_pagos  pax_pases_pagos  pax_franq  total  periodo  
0  Castro Barros        1.0              0.0        0.0    1.0   201801  
1           Lima        4.0              0.0        0.0    4.0   201801  
2          Pasco        1.0              0.0        0.0    1.0   201801  
3           P